src/libsyntax/parse/parser.rs

```diff
@@ -2266,6 +2266,26 @@ impl<'a> Parser<'a> {
         let (fieldname, expr, is_shorthand) = if self.look_ahead(1, |t| t == &token::Colon) {
             let fieldname = self.parse_field_name()?;
             self.bump(); // `:`
+            (fieldname, self.parse_expr()?, false)
+        // Check if a equals exists on ahead. This means the source incorrectly attempts to
+        // initialize a field with an eq rather than a colon.
+        } else if self.look_ahead(1, |t| t == &token::Eq) {
+            let fieldname = self.parse_field_name()?;
+
+            // Emit an error with a suggestion.
+            self.diagnostic()
+                .struct_span_err(self.span, "struct fields are initialized with a equals symbol")
+                .span_suggestion_with_applicability(
+                    fieldname.span.shrink_to_hi().to(self.span),
+                    "replace equals symbol with a colon",
+                    ":".to_string(),
+                    Applicability::MachineApplicable,
+                )
+                .emit();
+
+            // Move past the incorrect eq token.
+            self.bump();
+
```

Comment:
Nit: This `if` "arm" can be merged into the previous one, the only different thing it does is the extra error message.

Reply:
Oops, well spotted, fixed that.